File: python/x402_arc_sdk/adapters/langchain.py
from __future__ import annotations

from typing import Any, Optional

from pydantic import BaseModel, Field

from ..client import X402ArcClient


class PayResourceArgs(BaseModel):
    url: str
    label: Optional[str] = None
    max_usdc: Optional[str] = Field(default=None, pattern=r"^\d+(\.\d{1,6})?$")


class PayBatchArgs(BaseModel):
    urls: list[str] = Field(min_length=1, max_length=50)
    max_usdc_per_resource: Optional[str] = Field(default=None, pattern=r"^\d+(\.\d{1,6})?$")
    concurrency: int = Field(default=4, ge=1, le=16)


class DepositArgs(BaseModel):
    amount_usdc: str = Field(pattern=r"^\d+(\.\d{1,6})?$")
    approve_amount_usdc: Optional[str] = Field(default=None, pattern=r"^\d+(\.\d{1,6})?$")
    skip_approval: bool = False
    assume_approval_confirmed: bool = False
    idempotency_prefix: Optional[str] = None


def get_langchain_tools(client: X402ArcClient) -> list[Any]:
    try:
        from langchain_core.tools import StructuredTool
    except ImportError as exc:
        raise ImportError("Install optional dependency: pip install 'x402-arc-sdk[langchain]'") from exc

    return [
        StructuredTool.from_function(lambda url, label=None, max_usdc=None: str(client.pay_resource(url=url, label=label, max_usdc=max_usdc)), name="x402_pay_resource", description="Pay one x402 Circle Gateway resource using native Python + Circle DCW signTypedData. No raw buyer private key.", args_schema=PayResourceArgs),
        StructuredTool.from_function(lambda urls, max_usdc_per_resource=None, concurrency=4: str(client.pay_batch(urls=urls, max_usdc_per_resource=max_usdc_per_resource, concurrency=concurrency)), name="x402_pay_batch", description="Pay multiple x402 resources after fail-closed preflight.", args_schema=PayBatchArgs),
        StructuredTool.from_function(lambda: str(client.gateway_balance()), name="x402_gateway_balance", description="Check Circle Gateway USDC balance."),
        StructuredTool.from_function(lambda amount_usdc, approve_amount_usdc=None, skip_approval=False, assume_approval_confirmed=False, idempotency_prefix=None: str(client.gateway_deposit(amount_usdc=amount_usdc, approve_amount_usdc=approve_amount_usdc, skip_approval=skip_approval, assume_approval_confirmed=assume_approval_confirmed, idempotency_prefix=idempotency_prefix)), name="x402_gateway_deposit", description="Create Circle DCW approval/deposit contract execution transactions.", args_schema=DepositArgs),
        StructuredTool.from_function(lambda url: str(client.supports(url=url)), name="x402_supports", description="Probe whether a URL supports Circle Gateway x402 batching."),
    ]

File: src/adapters/langchain.ts
import { z } from "zod";
import type { BuyerBatchAgent } from "../core/buyer.js";

async function optionalImport<T = any>(specifier: string): Promise<T> {
  const dynamicImport = new Function("specifier", "return import(specifier)") as (specifier: string) => Promise<T>;
  return dynamicImport(specifier);
}

export async function getLangChainBuyerTools(buyer: BuyerBatchAgent): Promise<any[]> {
  let langchainTools: { tool: any };
  try {
    langchainTools = await optionalImport("@langchain/core/tools");
  } catch {
    throw new Error("Install optional peer dependency first: npm install @langchain/core");
  }
  const { tool } = langchainTools;

  const payResource = tool(
    async ({ url, label, maxUsdcOverride }: { url: string; label?: string; maxUsdcOverride?: string }) => JSON.stringify(await buyer.payResource({ url, label, maxUsdcOverride })),
    {
      name: "x402_pay_resource",
      description: "Pay one Circle Gateway x402-batched resource on Arc using Circle DCW. Fails closed if policy, budget, allowlist, balance, or Gateway support checks fail.",
      schema: z.object({
        url: z.string().url(),
        label: z.string().optional(),
        maxUsdcOverride: z.string().regex(/^\d+(\.\d{1,6})?$/).optional(),
      }),
    },
  );

  const payBatch = tool(
    async ({ urls, maxUsdcPerResource }: { urls: string[]; maxUsdcPerResource?: string }) => JSON.stringify(await buyer.payBatch({ requests: urls.map((url) => ({ url, maxUsdcOverride: maxUsdcPerResource })) })),
    {
      name: "x402_pay_batch",
      description: "Pay multiple x402 resources through Circle Gateway batched nanopayments. Performs preflight before sending payment signatures.",
      schema: z.object({ urls: z.array(z.string().url()).min(1).max(50), maxUsdcPerResource: z.string().regex(/^\d+(\.\d{1,6})?$/).optional() }),
    },
  );

  const balance = tool(
    async () => JSON.stringify(await buyer.getBalances()),
    {
      name: "x402_gateway_balance",
      description: "Check buyer DCW wallet and Circle Gateway USDC balances.",
      schema: z.object({}),
    },
  );


  const deposit = tool(
    async ({ amountUsdc, skipApproval, assumeApprovalConfirmed }: { amountUsdc: string; skipApproval?: boolean; assumeApprovalConfirmed?: boolean }) => JSON.stringify(await buyer.createGatewayDeposit(amountUsdc, { skipApproval, assumeApprovalConfirmed })),
    {
      name: "x402_gateway_deposit",
      description: "Create Circle DCW approval/deposit transactions to fund Gateway balance.",
      schema: z.object({
        amountUsdc: z.string().regex(/^\d+(\.\d{1,6})?$/),
        skipApproval: z.boolean().optional(),
        assumeApprovalConfirmed: z.boolean().optional(),
      }),
    },
  );

  const supports = tool(
    async ({ url }: { url: string }) => JSON.stringify(await buyer.supports(url)),
    {
      name: "x402_supports",
      description: "Probe whether a URL supports Circle Gateway x402 batching before paying.",
      schema: z.object({ url: z.string().url() }),
    },
  );

  return [payResource, payBatch, balance, deposit, supports];
}
